### PR TITLE
Django ORM

### PR DESCRIPTION
**Pull request by angelala3252**
In this PR, I added information about the Django ORM to the existing file on Django Rest Framework. In particular, I wrote an introduction to the Django ORM, how to define a model, how to make and apply migrations, and how to query the database.

### CODE REVIEW

File: Topics/Tech_Stacks/Django_Rest.md [hunk 1]
```diff
@@ -47,12 +47,17 @@ Django Rest Framework is a popular library for building RESTful APIs with Django
...
+ ### Django ORM
+ As Django is a backend framework, there is a need for persistent storage of any data used by the web application. Django supports relational databases through its own built-in ORM. An [ORM](https://www.freecodecamp.org/news/what-is-an-orm-the-meaning-of-object-relational-mapping-database-tools/) (Object Relational Mapper) provides a layer of abstraction from the database, allowing the user to define relations and make queries using objects, rather than having to write raw SQL code. This makes ORMs suitable for working with object-oriented programming. Another benefit of ORMs is that the application is separated from the actual database implementation, which means the database can be changed (for example, from SQLite to PostgreSQL once the app is ready for production) without affecting the app.
```
Comment (dianzrong) [inline Topics/Tech_Stacks/Django_Rest.md:54]: Could you rephrase the first line? I find it hard to understand.

File: Topics/Tech_Stacks/Django_Rest.md [hunk 2]
```diff
@@ -47,12 +47,17 @@ Django Rest Framework is a popular library for building RESTful APIs with Django
...
+
+ ### Defining a model
+
+Using the Django ORM, we can define models in the database by creating Python classes. Each table is represented by a class, with the columns represented by attributes. Django provides field types for various SQL data types. Django also comes with some predefined models, such as the [User model](https://docs.djangoproject.com/en/4.2/topics/auth/default/#user-objects) with built-in authentication (but we won't be using the User model in this tutorial). 
+
+To create our API, we must first create a new model to represent our API resource. To illustrate the use of Django Rest Framework, we'll create a simple model for a to-do item with two fields: its title as a string, and its completed status as a boolean value. We first create a new file called models.py in the API app folder and define our model:
```
Comment (dianzrong) [inline Topics/Tech_Stacks/Django_Rest.md:60]: Those are helpful edits!!

File: Topics/Tech_Stacks/Django_Rest.md [hunk 3]
```diff
@@ -62,6 +67,24 @@ class TodoItem(models.Model):
     completed = models.BooleanField(default=False)
 ```
 
+For this model, we have used a CharField for the string title. The CharField is used for small amounts of text, with a default max length of 255 characters. We have defined the max length to be 200 characters in this case. We have also used a BooleanField for the boolean completion status, and we have explicitly defined the default value to be False. Thus, whenever a value isn't provided for the completed field, it will automatically be set to False.
```
Comment (dianzrong) [inline Topics/Tech_Stacks/Django_Rest.md:70]: I think the second sentence should start with "A CharField"

File: Topics/Tech_Stacks/Django_Rest.md [hunk 4]
```diff
@@ -62,6 +67,24 @@ class TodoItem(models.Model):
...
+
+Now that we have created a new model, we must create a migration for these changes to reflect in the database.
+
+### Migrations
+A migration is a file that tells the ORM what changes have been made to the models. Whenever we make changes to our models, we must run the following two commands:
+- ```python3 manage.py makemigrations```: This command creates a migration file containing all the changes made since the last migration, similar to a Git history. These changes are not auto-applied as changing the database structure is an important (and potentially dangerous) operation.
+- ```python3 manage.py migrate```: This command applies changes from any unapplied migration files to the database using SQL code. This syncs the state of the database tables with the models defined as Python classes.
+
+Be VERY careful with your migrations. If you are ever unsure, consult the [documentation](https://docs.djangoproject.com/en/4.2/topics/migrations/)!
```
Comment (dianzrong) [inline Topics/Tech_Stacks/Django_Rest.md:81]: I think it would be good at add why students should be careful with migrations

File: Topics/Tech_Stacks/Django_Rest.md [hunk 5]
```diff
@@ -109,7 +132,7 @@ urlpatterns = [
 ]
 ```
 
-All in all, we have now built a simple API with the ability to list, create, update and delete todo list items. One can test the API by starting the Django development server using ``` python3 manage.py runserver ``` and visiting http://localhost:8000/todos/ in a web browser.
+All in all, we have now built a simple API with the ability to list, create, update and delete todo list items. One can test the API by starting the Django development server using ``` python3 manage.py runserver ``` and visiting http://localhost:8000/todos/ in a web browser. Running this command will also show any errors that need to be fixed in the shell, and this continuously updates whenever the code is changed.
```
Comment (dianzrong) [inline Topics/Tech_Stacks/Django_Rest.md:135]: Can you change the last line to indicate what continuously updates? I'm not sure what "this" is.

**Review by dianzrong — COMMENTED** [review]
This document was really well written and helped me better understand Django ORMs. I just have some small comments.

File: Topics/Tech_Stacks/Django_Rest.md (hunk 5 above)
Comment (angelala3252) [reply Topics/Tech_Stacks/Django_Rest.md]: [ea08aee](https://github.com/learning-software-engineering/learning-software-engineering.github.io/pull/310/commits/ea08aee1dd22045fff731064ae7c678a898aa516)

File: Topics/Tech_Stacks/Django_Rest.md (hunk 4 above)
Comment (angelala3252) [reply Topics/Tech_Stacks/Django_Rest.md]: [ea08aee](https://github.com/learning-software-engineering/learning-software-engineering.github.io/pull/310/commits/ea08aee1dd22045fff731064ae7c678a898aa516)

File: Topics/Tech_Stacks/Django_Rest.md (hunk 3 above)
Comment (angelala3252) [reply Topics/Tech_Stacks/Django_Rest.md]: [ea08aee](https://github.com/learning-software-engineering/learning-software-engineering.github.io/pull/310/commits/ea08aee1dd22045fff731064ae7c678a898aa516)

File: Topics/Tech_Stacks/Django_Rest.md (hunk 1 above)
Comment (angelala3252) [reply Topics/Tech_Stacks/Django_Rest.md]: [0dc54a6](https://github.com/learning-software-engineering/learning-software-engineering.github.io/pull/310/commits/0dc54a645881d5e655fbf217ba2a9bd386733247)

File: Topics/Tech_Stacks/Django_Rest.md [hunk 6]
```diff
@@ -47,12 +47,17 @@ Django Rest Framework is a popular library for building RESTful APIs with Django
...
+
+ ### Defining a model
+
+Using the Django ORM, we can define models in the database by creating Python classes. Each table is represented by a class, with the columns represented by attributes. Django provides field types for various SQL data types. Django also comes with some predefined models, such as the [User model](https://docs.djangoproject.com/en/4.2/topics/auth/default/#user-objects) with built-in authentication (but we won't be using the User model in this tutorial).
```
[the remark this answers is not in the document]
Comment (angelala3252) [reply Topics/Tech_Stacks/Django_Rest.md]: [0dc54a6](https://github.com/learning-software-engineering/learning-software-engineering.github.io/pull/310/commits/0dc54a645881d5e655fbf217ba2a9bd386733247)

**Comment by angelala3252** [comment]
@dianzrong I've made all your requested changes, please take a look and let me know what you think!

File: Topics/Tech_Stacks/Django_Rest.md [hunk 7]
```diff
@@ -67,18 +67,18 @@ class TodoItem(models.Model):
...
 
 Now that we have created a new model, we must create a migration for these changes to reflect in the database.
 
 ### Migrations
 A migration is a file that tells the ORM what changes have been made to the models. Whenever we make changes to our models, we must run the following two commands:
-- ```python3 manage.py makemigrations```: This command creates a migration file containing all the changes made since the last migration, similar to a Git history. These changes are not auto-applied as changing the database structure is an important (and potentially dangerous) operation.
+- ```python3 manage.py makemigrations```: This command creates a migration file containing all the changes made since the last migration, similar to a Git history. These changes are not auto-applied as changing the database structure is an important (and potentially dangerous) operation that can result in data loss.
```
Comment (dianzrong) [inline Topics/Tech_Stacks/Django_Rest.md:80]: I love the specifics!

File: Topics/Tech_Stacks/Django_Rest.md [hunk 8]
```diff
@@ -67,18 +67,18 @@ class TodoItem(models.Model):
...
 - ```python3 manage.py migrate```: This command applies changes from any unapplied migration files to the database using SQL code. This syncs the state of the database tables with the models defined as Python classes.
 
-Be VERY careful with your migrations. If you are ever unsure, consult the [documentation](https://docs.djangoproject.com/en/4.2/topics/migrations/)!  
+Be VERY careful with your migrations. If you accidentally delete a migration file, edit the migration file by hand, or edit the same model as someone else at  the same time, you could create migration errors that are very complicated to resolve. If you are ever unsure, consult the [documentation](https://docs.djangoproject.com/en/4.2/topics/migrations/)!
```
Comment (dianzrong) [inline Topics/Tech_Stacks/Django_Rest.md:83]: Tiny thing: there are 2 spaces between "at" and "the"

**Review by dianzrong — COMMENTED** [review]
One tiny thing, but other than that, I love the changes!

File: Topics/Tech_Stacks/Django_Rest.md (hunk 8 above)
Comment (angelala3252) [reply Topics/Tech_Stacks/Django_Rest.md]: [3d25496](https://github.com/learning-software-engineering/learning-software-engineering.github.io/pull/310/commits/3d25496173e1cce072478c9e2ca37a7b0a6c0db1)

**Comment by dianzrong** [comment]
Looks amazing and so ready to merge!!
